File: Python/Find the City With the Smallest Number of Neighbors at a Threshold Distance.py
from sys import maxsize
from typing import List


class Solution:
    """
    Time:   O(n^3)
    Memory: O(n^2)
    """

    def findTheCity(self, n: int, edges: List[List[int]], distance_threshold: int) -> int:
        dist = [[maxsize] * n for _ in range(n)]

        for u, v, w in edges:
            dist[u][v] = dist[v][u] = w

        for i in range(n):
            dist[i][i] = 0

        for k in range(n):
            for i in range(n):
                for j in range(n):
                    dist[i][j] = min(dist[i][k] + dist[k][j], dist[i][j])

        min_city = min_neighbours = maxsize
        for i in range(n):
            neighbours = sum(dist[i][j] <= distance_threshold for j in range(n))
            if neighbours <= min_neighbours:
                min_city = i
                min_neighbours = neighbours

        return min_city

File: Java/Find the City With the Smallest Number of Neighbors at a Threshold Distance.java
public class Solution {

    /**
     * Time:   O(n^3)
     * Memory: O(n^2)
     */
    public int findTheCity(int n, int[][] edges, int distanceThreshold) {
        int[][] dist = new int[n][n];

        for (int i = 0; i < n; ++i)
            for (int j = 0; j < n; ++j)
                if (i != j)
                    dist[i][j] = INF;

        for (int[] edge : edges) {
            dist[edge[0]][edge[1]] = edge[2];
            dist[edge[1]][edge[0]] = edge[2];
        }

        for (int k = 0; k < n; ++k)
            for (int i = 0; i < n; ++i)
                for (int j = 0; j < n; ++j)
                    dist[i][j] = Math.min(dist[i][k] + dist[k][j], dist[i][j]);

        int minCity = 0, minNeighbours = n;
        for (int i = 0; i < n; ++i) {
            int neighbours = 0;

            for (int j = 0; j < n; ++j)
                if (dist[i][j] <= distanceThreshold) neighbours += 1;

            if (neighbours <= minNeighbours) {
                minCity = i;
                minNeighbours = neighbours;
            }
        }

        return minCity;
    }

    public static final int INF = 1_000_000;
}
